rotors_gazebo_plugins/src/gazebo_multimotor_plugin.cpp

```diff
@@ -0,0 +1,217 @@
+
+  // Initialise with default namespace (typically /gazebo/default/)
+  node_handle_->Init();
+
+  getSdfParam<std::string>(_sdf, "actuatorCommandSubTopic", command_actuator_sub_topic_,
+                           command_actuator_sub_topic_);
+  getSdfParam<std::string>(_sdf, "actuatorStatePubTopic", motor_state_pub_topic_,
+                           motor_state_pub_topic_);
+
+  //=============================================//
+  //========== LOAD ROTORS AND SERVOS ===========//
+  //==============================================//
+  std::string joint_name, link_name;
+
+  // Add rotors.
+  if (_sdf->HasElement("rotors")) {
+    sdf::ElementPtr motors = _sdf->GetElement("rotors");
+    sdf::ElementPtr motor = motors->GetElement("rotor");
+
+    while (motor) {
+      // Only load valid motors
+      if (IsValidLink(motor) & IsValidJoint(motor)) {
+        motors_.push_back(std::make_unique<MotorModelRotor>(model_, motor));
+        gzdbg << "[gazebo_multimotor_plugin] Loaded rotor!\n";
+      } else {
+        gzdbg << "[gazebo_multimotor_plugin] Failed to load rotor!\n";
+      }
+      motor = motor->GetNextElement("rotor");
+    }
+  }
+
+  // Add servos.
+  if (_sdf->HasElement("servos")) {
+    sdf::ElementPtr motors = _sdf->GetElement("servos");
+    sdf::ElementPtr motor = motors->GetElement("servo");
+
+    while (motor) {
+      // Only load valid motors
+      if (IsValidJoint(motor)) {
+        motors_.push_back(std::make_unique<MotorModelServo>(model_, motor));
+        gzdbg << "[gazebo_multimotor_plugin] Loaded servo!\n";
+      } else {
+        gzdbg << "[gazebo_multimotor_plugin] Failed to load servo!\n";
+      }
+      motor = motor->GetNextElement("servo");
+    }
+  }
+  gzdbg << "[gazebo_multimotor_plugin] Loaded " << motors_.size() << " actuators.";
+
+  // Listen to the update event. This event is broadcast every
+  // simulation iteration.
+  updateConnection_ = event::Events::ConnectWorldUpdateBegin(
+      boost::bind(&GazeboMultimotorPlugin::OnUpdate, this, _1));
+}
+
+void GazeboMultimotorPlugin::OnUpdate(const common::UpdateInfo&) {
+  if (!pubs_and_subs_created_) {
+    CreatePubsAndSubs();
+    pubs_and_subs_created_ = true;
+  }
+
+  gz_sensor_msgs::Actuators actuator_state_msg;
+  double position, velocity, effort;
+
+  for (const auto& motor : motors_) {
+    motor->UpdatePhysics();
+    motor->GetActuatorState(&position, &velocity, &effort);
+    actuator_state_msg.add_angles(position);
+    actuator_state_msg.add_angular_velocities(velocity);
+    actuator_state_msg.add_normalized(effort);
+  }
+
+  common::Time now = world_->SimTime();
+
+  actuator_state_msg.mutable_header()->mutable_stamp()->set_sec(now.sec);
+  actuator_state_msg.mutable_header()->mutable_stamp()->set_nsec(now.nsec);
+  // Frame ID is not used for this particular message
+  actuator_state_msg.mutable_header()->set_frame_id("");
+
+  motor_state_pub_->Publish(actuator_state_msg);
```

Comment:
Current publishing on simulation update, which is not realistic. Should probably align with the real system. We could:

- poll in the command function, but then we wouldn't see feedback when no commands sent.
- poll based on a timed callback

What do you think @clanegge @michaelpantic ?